### PR TITLE
Voice preloading

### PR DESCRIPTION
**Pull request by styagi130**
Title: Voice Preloading

**Source branch:** `voice-preloading`  
**Target branch:** `main`  
**Head:** `d2a7273656ce9314e76929273d24e22277e2ab6a`  
**Comparison base:** local `main` at `ea1ebf55b25fceff372f885b4e7b5803f3d33dea`.

# What does this PR do ?

Load known EasyMagpie voices during startup so a context-cache miss can reuse validated speaker data. Keep model embeddings and API prompt-length calculations consistent, with lazy loading for voices that were skipped, added or replaced after startup.

**Collection**: TTS (`easymagpie_vllm_omni`).

## Scope against main

This is MR 3 of 5. Its branch includes the following prerequisite changes in the current diff to `main`:

- **Serving & Benchmarks** (`serving-benchmarks`).
- **Decoding & Audio Transfers** (`decoding-audio-transfers`).

The changelog below describes the additional work in this layer. The full current comparison is **10 commits; 47 files changed, 4876 insertions(+), 440 deletions(-)**. Merge the preceding MRs first. If their commits are preserved, those changes drop out of this diff as `main` advances. After a squash or rebase, restack and revalidate this branch before merging.

# Changelog

- Share a validated loader and file fingerprints between the model and serving adapter for supported tensor/dictionary `.pt` voices.
- Validate tensor shape, layout, finite values and dtype conversion before retaining a voice.
- Inspect at most 64 sorted entries, skip files larger than 16 MiB, and cap retained startup buffers at 16 MiB per language-model worker. The API retains frame counts only.
- Reuse unchanged preloaded voices on context-cache misses while preserving existing cached-context behavior and lazy fallback for skipped, late or replaced files.
- Add tests for validation, bounds, ownership and preloaded/lazy equivalence, and document the supported input forms and startup limits.

# Usage

Continue placing voice files at `speaker_embeddings/<voice>.pt`. Each file may be a 2-D tensor or a dictionary containing `speaker_encoding`. Preloading runs automatically; there is no registration step or new client field. Retention limits bound stored startup buffers, not peak deserialization memory.

# GitHub Actions CI

Remote CI is pending. Trusted PRs use the existing copy-pr-bot flow. For an untrusted PR, a maintainer can trigger CI with `/ok to test <head-sha>` after publication.

## Validation

Recorded at `d2a7273656ce9314e76929273d24e22277e2ab6a` on **2026-09-15**: **529 tests passed, zero failures and zero skips** in the complete CPU/CUDA serving suite. Tests used physical **H100 device 1**, exposed as the sole logical CUDA device in the container.

The package was freshly installed from this head. Source hashes, installed module origins, `pip check`, dependency invariance and author-matching DCO sign-offs passed. Full-branch pre-commit passed from the recorded main base. The runtime used vLLM/vLLM-Omni 0.26.0, PyTorch 2.11.0+cu130 and Triton 3.6.0.

<details>
<summary>Recorded validation commands</summary>

The serving command ran inside the pinned test container after installing this branch and providing the codec contract fixture:

```bash
/usr/bin/python3 -I -B /recipe/installed_pytest.py \
  --rootdir=/snapshot \
  --confcutdir=/snapshot/tests/collections/tts/easymagpie_vllm_omni/serving \
  /snapshot/tests/collections/tts/easymagpie_vllm_omni/serving \
  -p no:cacheprovider -q -ra --junitxml=/results/tests.xml \
  --durations=10
```

Full-branch formatting checks and the current diff check:

```bash
python3 -m pre_commit run \
  --from-ref ea1ebf55b25fceff372f885b4e7b5803f3d33dea \
  --to-ref d2a7273656ce9314e76929273d24e22277e2ab6a
git diff --check ea1ebf55b25fceff372f885b4e7b5803f3d33dea d2a7273656ce9314e76929273d24e22277e2ab6a
```

</details>

Live HTTP/WebSocket audio serving, cancellation/recovery, shutdown/drain, full-utterance quality/completion, performance acceptance and remote CI remain pending. The local test counts do not establish benchmark throughput or latency.

# Before your PR is "Ready for review"

- [x] Add regression tests and documentation with the implementation.
- [x] Pass the complete installed CPU/CUDA serving suite at the recorded head.
- [x] Pass full-branch pre-commit, diff checks and author-matching DCO checks.
- [ ] Complete contributor-guideline review and reviewer assessment of tests, documentation and optional dependency guards.
- [ ] Complete the remaining live serving, quality/completion and performance checks; attach portable validation evidence and pass remote CI.

**PR Type**: New Feature.

## Who can review?

TTS serving maintainers familiar with speaker inputs and context caching.

# Additional Information

Related to NVIDIA-NeMo/Speech#16220. Merge order: `serving-benchmarks` → `decoding-audio-transfers` → `voice-preloading` → `input-preparation-routing` → `h100-deployment`.

### CODE REVIEW

File: tools/easymagpie_vllm_omni/easymagpie_vllm_omni/speakers.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+"""Bounded, optional startup loading for the existing known-voice .pt format."""
+
+import logging
+import pickle
+import struct
+from pathlib import Path
+
+import torch
+
+MAX_PRELOAD_ENTRIES = 64
```
Comment (vklimkov-nvidia) [inline tools/easymagpie_vllm_omni/easymagpie_vllm_omni/speakers.py:23]: bad default. we will have more

File: tools/easymagpie_vllm_omni/easymagpie_vllm_omni/speakers.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+MAX_PRELOAD_BYTES = 16 * 1024 * 1024
```
Comment (vklimkov-nvidia) [inline tools/easymagpie_vllm_omni/easymagpie_vllm_omni/speakers.py:24]: very arbitrary. why do we have this limits at all?

File: tools/easymagpie_vllm_omni/easymagpie_vllm_omni/easymagpie.py [hunk 3]
```diff
@@ -849,35 +843,23 @@ def _resolve_speaker_embedding(self, device: torch.device, info_dict: dict[str,
...
-        via :meth:`load_weights`) so known speakers work even under
-        ``--load-format dummy``, which skips weight loading.
-        """
+        """Reuse an unchanged startup voice; keep late/replaced voices lazy."""
         import glob
         import os
 
         spk_dir = os.path.join(self.model_path, "speaker_embeddings")
         path = os.path.join(spk_dir, f"{speaker_id}.pt")
+        prepared = getattr(self, "_speaker_buffers", {}).get(speaker_id)
```
Comment (vklimkov-nvidia) [inline tools/easymagpie_vllm_omni/easymagpie_vllm_omni/easymagpie.py:852]: no real reason to use getattr here

File: tools/easymagpie_vllm_omni/easymagpie_vllm_omni/speakers.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        raise ValueError(f"Skipped speaker embedding {path}: startup storage budget exceeded")
+    if not torch.isfinite(embedding).all().item():
+        raise ValueError(f"Invalid speaker embedding {path}: non-finite values")
+    embedding = embedding.detach().to(device="cpu", dtype=dtype, copy=True, memory_format=torch.contiguous_format)
+    if not torch.isfinite(embedding).all().item():
+        raise ValueError(f"Invalid speaker embedding {path}: values overflow {dtype}")
+    return embedding
+
+
+def speaker_fingerprint(path):
```
Comment (vklimkov-nvidia) [inline tools/easymagpie_vllm_omni/easymagpie_vllm_omni/speakers.py:67]: why do we need fingerprinting at all? seems like an akward attempt to support hot change of voices for the service that is running. but i dont think we want to support such functionality yet. when we do, that requires separate change. current fingerprinting doesnt quite work anyway: existing assembled contexts retain old voice, model might use different speaker embedding for same speaker_id depending on context_text.

i'd drop it alltogether.

File: tools/easymagpie_vllm_omni/easymagpie_vllm_omni/speakers.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+        OSError,
+        EOFError,
+        RuntimeError,
+        ValueError,
+        IndexError,
+        KeyError,
+        pickle.UnpicklingError,
+        struct.error,
+    ) as error:
+        raise ValueError(f"Cannot load speaker embedding {path}: {error}") from error
```
Comment (vklimkov-nvidia) [inline tools/easymagpie_vllm_omni/easymagpie_vllm_omni/speakers.py:43]: what exactly is the point of catching error?

File: tools/easymagpie_vllm_omni/easymagpie_vllm_omni/speakers.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    embedding = loaded.get("speaker_encoding") if isinstance(loaded, dict) else loaded
+    if (
+        type(embedding) is not torch.Tensor
+        or embedding.device.type != "cpu"
+        or embedding.layout != torch.strided
+        or embedding.is_quantized
+        or embedding.is_complex()
```
Comment (vklimkov-nvidia) [inline tools/easymagpie_vllm_omni/easymagpie_vllm_omni/speakers.py:50]: why would that happen? why do we need all these checks?

File: tools/easymagpie_vllm_omni/easymagpie_vllm_omni/speakers.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+        stat = Path(path).stat()
+    except OSError:
+        return None
+    return stat.st_dev, stat.st_ino, stat.st_size, stat.st_mtime_ns
+
+
+def preload_speakers(model_path, *, width=None, dtype=torch.float32):
+    """Yield valid startup voices; skipped/late voices retain lazy request loading."""
+    try:
+        paths = sorted((Path(model_path) / "speaker_embeddings").glob("*.pt"))[:MAX_PRELOAD_ENTRIES]
```
Comment (vklimkov-nvidia) [inline tools/easymagpie_vllm_omni/easymagpie_vllm_omni/speakers.py:79]: why do we sort?

File: tools/easymagpie_vllm_omni/README.md [hunk 8]
```diff
@@ -85,6 +85,14 @@ Container caches and uploaded speaker samples default to writable `/tmp` paths.
 Override the cache variables with writable persistent mounts when reusing compilation caches.
 The launcher uses `exec` for signal forwarding; graceful shutdown still needs runtime validation.
 
+Known voices remain `speaker_embeddings/<voice>.pt`: a 2-D tensor or a dictionary's `speaker_encoding` tensor.
```
Comment (vklimkov-nvidia) [inline tools/easymagpie_vllm_omni/README.md:88]: drop these impl details from top level README

**Review by vklimkov-nvidia — COMMENTED** [review]
overall a good idea to have spkr embeddings as model buffer.

certain checks/workarounds are not prompted and actually harmful.

there is duplication in calculating prompt length. lets keep it in a single place.

no need to support 2 formats of the speaker embeddings, lets use a single one (dict).

adapter can just use get_prompt_len from model. we dont care if there is slight overhead on the very first synthesis. in prod it doesnt matter, in benchmark we do warmup.
Therefore new adapter pieces can be dropped:
- _speaker_lengths
- _preload_speaker_lengths()
- Adapter-side preload_speakers()
- Adapter-side fingerprint checks
- Manually duplicated task_len + speaker_len + context_len calculation
- Corresponding adapter preload/replacement/budget tests

we need to check if we actually gonna use `context_text` and if its going to change at all for the given voice. if not, then the reason for this change disappears altogether. we cache the entire embedding (task, spkr, context_text) and that is enough performance wise.
